packages/jsii/lib/compiler.ts

```diff
@@ -495,23 +497,43 @@ export class Compiler implements Emitter {
    * This makes it so that running 'tsc' and running 'jsii' has the same behavior.
    */
   private determineSources(files: string[]): string[] {
-    const ret = new Array<string>();
+    const ret = new Set<string>();
 
     if (files.length > 0) {
-      ret.push(...files);
+      for (const file of files) {
+        ret.add(file);
+      }
     } else {
       const parseConfigHost = parseConfigHostFromCompilerHost(
         this.compilerHost,
       );
-      const parsed = ts.parseJsonConfigFileContent(
+      const { fileNames } = ts.parseJsonConfigFileContent(
         this.typescriptConfig,
         parseConfigHost,
         this.options.projectInfo.projectRoot,
       );
-      ret.push(...parsed.fileNames);
+      for (const file of fileNames) {
+        ret.add(file);
+      }
     }
 
-    return ret;
+    // Bonus: ensure all dependencies' entry points are included in the compiler
+    // input path. This guarantees we have symbols for all types, from the
+    // module root, which is necessary in order to properly detect submodules.
+    for (const assm of this.options.projectInfo.dependencyClosure) {
```

Comment:
Do we really need the closure? Would the direct dependencies not be good enough?

Reply:
Yes it is. We need to do the submodule traversal for _all_ transitive dependencies, as types thereof might be referred to implicitly but the user code (i.e: via a forwarded return value), but explicitly in the type model.

The problem we have where the submodule traders could not work if the dependency's entry point is not in the compiler's input set also applies to transitive dependencies.